java/simplesaga-model/src/main/java/io/simplesource/saga/model/messages/SagaStateTransition.java

```diff
@@ -10,30 +10,63 @@
 import java.util.List;
 import java.util.Optional;
 import java.util.UUID;
+import java.util.function.Function;
 
 public interface SagaStateTransition {
+
     @Value
-    class SetInitialState<A> implements SagaStateTransition {
+    final class SetInitialState<A> implements SagaStateTransition {
         public final Saga<A> sagaState;
+
+        @Override
+        public <A> A cata(Function<SetInitialState<?>, A> f1, Function<SagaActionStatusChanged, A> f2, Function<SagaStatusChanged, A> f3, Function<TransitionList, A> f4) {
+            return f1.apply(this);
+        }
     }
 
     @Value
-    class SagaActionStatusChanged implements SagaStateTransition {
+    final class SagaActionStatusChanged implements SagaStateTransition {
         public final UUID sagaId;
         public final UUID actionId;
         public final ActionStatus actionStatus;
         public final Optional<SagaError> actionError;
+
+        @Override
+        public <A> A cata(Function<SetInitialState<?>, A> f1, Function<SagaActionStatusChanged, A> f2, Function<SagaStatusChanged, A> f3, Function<TransitionList, A> f4) {
+            return f2.apply(this);
+        }
     }
 
     @Value
-    class SagaStatusChanged implements SagaStateTransition {
+    final class SagaStatusChanged implements SagaStateTransition {
         public final UUID sagaId;
         public final SagaStatus sagaStatus;
         public final Optional<NonEmptyList<SagaError>> actionErrors;
+
+
+        @Override
+        public <A> A cata(Function<SetInitialState<?>, A> f1, Function<SagaActionStatusChanged, A> f2, Function<SagaStatusChanged, A> f3, Function<TransitionList, A> f4) {
+            return f3.apply(this);
+        }
     }
 
     @Value
-    class TransitionList implements SagaStateTransition {
+    final class TransitionList implements SagaStateTransition {
         public final List<SagaActionStatusChanged> actions;
+
+        @Override
+        public <A> A cata(Function<SetInitialState<?>, A> f1, Function<SagaActionStatusChanged, A> f2, Function<SagaStatusChanged, A> f3, Function<TransitionList, A> f4) {
+            return f4.apply(this);
+        }
     }
+
+    /**
+     * Catamorphism over SagaStateTransition
+     */
+    <A> A cata(
```

Comment:
Happy with the pattern. Nice to remove the `instanceof`s.

Would you object to 
```
<A> Saga<A> cata(
            Function<SetInitialState<?>, Saga<A>> f1,
            Function<SagaActionStatusChanged, Saga<A>> f2,
            Function<SagaStatusChanged, Saga<A>> f3,
            Function<TransitionList, Saga<A>> f4
            );
```
I know it's a bit wordier, and that specialisation is not necessary, but it's not like the `cata` method with that signature can be applied to anything else but a `Saga`.

If not, I'd prefer to change the `A` to `B`, `S` or `SA` or anything else.

`A` as a type parameter is already widely used as the type representing the payload of saga actions.

Reply:
I'll change it to a B as it's more generic and no reason to constrain to `Saga<A>`.